llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -267,7 +267,7 @@ bool hasDoubleType(const Type *T) {
   if (T->isDoubleTy())
     return true;
   for (const Type *TT : T->subtypes())
-    if (TT->isDoubleTy())
+    if (hasDoubleType(TT))
```

Comment:
I'm concerned about potential infinite recursion here. There is a similar function which collects aspects used by a type and it uses a cache to ensure that no type is visited twice.

I would recommend checking an example where you have something like this:
```c++
struct B;
struct A {
 B *b;
};
struct B {
  A *a;
};
```

Reply:
Hi @AlexeySachkov 

Thanks. That was a good catch.
I tried this locally and LLVM IR generated looks like this:
%struct.T1 = type { ptr addrspace(4) }
%struct.T2 = type { ptr addrspace(4) }
So, we are saved here by opaque pointer types.
I added a test for this
Thanks